\documentclass[12pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[margin=1in]{geometry}
\usepackage{helvet}
\usepackage{tikz}
\usepackage{multirow}
\usepackage{caption}
\usepackage{array}
\usepackage{subcaption}
\usepackage[section]{placeins}
\usepackage[acronym]{glossaries}
\usepackage{setspace}



\title{%
    mRNA Isoform Loci Prediction from RNA Sequencing Experiments \\
    \large Thesis Proposal}

\author{Andrei Stanciu}
\date{December 2020}


\makeglossaries

\doublespacing
\begin{document}

\maketitle

\section{Introduction}

\subsection{History}
Genetic sequencing has seen an astonishing development in the past century \cite{Hood2003}. Only in 1953, Watson and Crick \cite{WATSON1953} proposed the now well known double helix structure of DNA. Soon after, in 1977, Maxam and Gilbert \cite{Maxam1977} described a method of sequencing DNA. Back then, it took Sanger et al. \cite{Sanger1977} roughly one year to sequence the genome of a virus, consisting of 5,375 nucleotides. By 1986, Smith et al. \cite{Smith1986} showcased a tool that would sequence roughly 250 nucleotides an hour. 

At the same time, in the 1940s and the 1950s the precursors of modern day computers were being created. Since then, the computing power increased at an exponential rate, as predicted by Moore \cite{Moore2006}. Together with the computing speed also grew the DNA sequencing throughput. While the Human Genome Project took 10 years to complete, Illumina Sequencing \cite{illuminaSeq} was sequencing 18000 human genomes in 2014. This series of breakthroughs serve as the foundation of what is now the 'genomic era', where the average researcher has access to myriads of genomic data sets and have the power to process them on their personal computers.   

\subsection{Biological Background}
The DNA holds a series of properties that make it suitable for digital analysis. A DNA molecule consists of two chains of nucleotides, also called strands. In turn, a strand of DNA is a series of  base pairs - adenine (A), cytosine (C), guanine (G) and thymine (T), which could come in any combination. These sequences of basepairs encode instructions for making the molecules that support life. 


Due to its underlying chemistry, the strand of DNA has two distinct ends called the 5' end and the 3' end, being oriented from the 5' end to the 3'end. Additionally, each nucleotide from one strand pairs with a corresponding nucleotide from the opposite strand. Thus, adenine (A) pairs with thymine (T) and cytosine (C) pairs with guanine (G). This, together with the fact that the two strands of the helix run in opposite (anti parallel) directions, produces for every sequence of DNA a reverse complement sequence. 

Sequences of DNA are copied (transcribed) to RNA and transported outside the nucleus. The RNA is a single stranded chain of nucleotides where Thymine (T) is replaced by Uracil (U). If the RNA holds instructions for making protein it is called \gls{mrna}. Even if only a small fraction of the human DNA codes for proteins, it is now believed that more than 90\% of the human genome gets transcribed \cite{Pertea2012}. Transcription itself is a highly regulated cellular process, responding to various cellular and extracellular signals\cite{Hahn2011}. This gives that the number of \acrshort{mrna}s inside the cell varies significantly from gene to gene, and from cell to cell. The collection of all mRNAs in a cell, called a transcriptome, define the identity of a cell in multi cellular organisms. 

\subsection{Gene Translation}
Inside the cell the \acrshort{mrna} gets translated into protein. This itself is a highly complex and highly regulated process. The basepairs in an \acrshort{mrna} are grouped in nonoverlapping sequences of 3 nucleotides, called codons. 61 of the possible codons code for a aminoacid, while 3 of them represent a stop codon. Ribosomes - macromomolecular machines consisting of ribosomal RNA and ribosomal proteins - bind to the mRNA and use the information in it to build the appropriate protein, aminoacid by aminoacid. A lot of the complexity and regulation of translation comes from the fact that the beginning and end of a protein are marked by a start and stop codon on the mRNA. Since the mRNA strands are oriented from their 5' end to their 3' end, the region before the start codon on an mRNA is called 5'\gls{utr} and the region following the stop codon is called 3'\acrshort{utr}. The region spanning  between the two codons, which codes for a protein, is  called \gls{utr}. Thus the ribosome buynd to the mRNA has to scan the 5'UTR until it identifies the AUG (start) codon. This codon signals the ribosome to start building the protein. 

\begin{enumerate}
    \item full details of ribosome translating? 
    \item Factors that influence translation
    \item An example of uORFs and how their expression affects the translation of the gene. [what was the canonical example for this?]
    \item significance of mRNA isoform variety in gene translation
\end{enumerate}

\subsection{RNASeq}
    Understanding and comparing the transcriptomes of different cells under different conditions offers sceintists a plethora of new insights. [give examples from papers]. With the wide availability of high throughput sequencing it is now possible to sequence all the RNA  present in an experiment\cite{Kukurba2015}. In the current technological era, such an mRNA experiment usually gives information about millions of basepairs present in a cell. 
    One common sequencing protocol is the Illumina sequencing. It excels at sequencing relatively short mRNA fragments (on the order of hundreds of nucleotides per fragment). [needs quotations] The way these experiments are prepared and sequenced involve multiple steps which each can indroduce a specific bias. In the case of Illumina sequencing, a biological sample often consists of multiple cells with the same condition. The cells in the sample are broken apart to expose the mRNAs and the mRNAs are filtered out. The purified mRNA subsequently gets broken up in smaller fragments. There are different methods available for fragmentation depending on the experimental design and desired outcomes. The fragmented mRNAs are then reverse transcribed to complementary DNA (cDNA), which is augumented with adapters and barcodes. [What is the purpose of the adapter?]. The barcodes serve as a unique identifier for the experiment. In the case multiple experiments are sequenced at the same time, the barcodes are used to tell which read comes from which experiment. Subsequently, a PCR [citation? details?] reaction amplifies the number of present fragments by making millions [?] of copies of each fragment. [What's the purpose of this? Is it to give us more confidence in the reads] [to be continued]   
    An additional variation in sequencing methodology comes from paired end versus single end reads. Single end reads are sequences of RNA read from one of the fragments described above. Based on the technology used [citation], for each fragment sequenced there is exactly one single end read. In the case of paired end reads, the RNA fragment is sequenced twice, once from each end. This gives two different reads. The sequencer imposes some limits on the length of the reads, and thus, the selected length of fragments influences the downstream possibilities of analysis. If the selected length is short enough, it is very likely that the two paired reads will overlap. This increases the confidence in each one of the reads. If the length of RNA is larger than twice the sequencing length, it becomes unlikely the two ends will overlap. This later option gives an ordering between pairs of reads and can be used for \textit{de novo} transcriptome assembly \cite{Hlzer2019}.  
\begin{enumerate}
    \item Quick intro on the biological preps: cells, mRNA isolation, fragmentation, PCR: note that these process introduce biases in which data gets sequenced
    \item general methods for sequencing
    \item paired end versus single end
    \item ribosome profiling
    \item our data
    \item Steinmetz RNASeq data
    \item note how much variety is in  RNASeq experiments and methods. 
\end{enumerate}

\subsection{mRNA isoforms in yeast}
 [ here there are many pepares I saved that I can cite and  talk about]



\subsection{Previous research/Literature review}
[papers on isoform analysis]

\subsection{mRNA Isoform  identification }
[Main goal of our tool]
\subsubsection{TODO list}

[Maybe use the figure from here as an example http://shadarf.blogspot.com/2017/07/how-to-make-reverse-complement-of-dna.html]

Introduce mRNA, 1 copy of DNA vs many copies of mRNA. 
Introduce Codons + Aminoacids 
Protein

\subsection{Introduction Outline}
\begin{enumerate}
    \item Introduction on the basic biology: From DNA to mRNA to protein
    \item Background on transcription 
    \item Introduce yeast as a model organism
    \item Genome annotations
    \item Introduce RNASequencing
    \item Present general methods for RNASeq
    \item Present Ingolia methods for RNASeq and introduce our datasets
    \item Introduce translation regulation. this is why we're trying to identify mRNA isoforms (in 5'UTR)
    \item Describe what isoforms are 
    \item Introduce the problem raised by steinmetz et el (new discoveries on yeast heterogeneity) 
    \item Talk about existing research on isoform analysis 
    \item Motivation behind using RNASeq to identify isoforms
    \item Present the main idea of the tool
    \item Present  the additional tools that resulted as a side effect from developing the main tool

\end{enumerate}

Punchline: \\

This study proposes a suite of software tools that use machine learning models to identify novel protein isoforms from \gls{rnaseq} Data. The suite also comes with a series of data processing and visualisation pipelines that are meant to help identifying new or miss annotated genes. In order to facilitate this, we provide a standalone in-silico data generation pipeline which produces novel RNASeq libraries informed by expected isoform distributions. 


\section{Proposal - Introduction}


Sequencing DNA has been one of the most impactful discoveries of the past century. With the recent advances \cite{Ingolia2012} in sequencing speed and availability, biologists are turning more and more to computational tools to analyze large data sets covering entire genomes. \gls{rnaseq} has become a widespread and accessible tool for scientists who are pushing the frontiers of science. [some intro aboyt the prep] After a library is prepared in the lab, all the \gls{mrna} present in the cells gets broken down into tinier fragments. \acrshort{rnaseq} reports with single nucleotide precision all these fragments. These fragments become digital reads represented as strings of A, C, G, T/U characters, and are subsequently aligned to the genome (digitally represented as large text files using the same encoding of nucleobases as above). The genome itself is divided in genes which are well annotated {\tiny[TODO: reference to annotations of the yeast genome]}. Each gene consists of three regions: 5’ \gls{utr}, \gls{cds} and 3’\acrshort{utr}. While the \acrshort{cds} is the region that gets translated into protein, the \acrshort{utr}s play an important role in regulation and translation. Many genes are known to have various isoforms - slight variation in length or in splicing - especially in more complex organisms. Understanding these variety of these isoforms could help us gain a more in depth understanding on regulatory mechanisms. For instance, a recent study \cite{Shamir2020} showed that 
the relative distribution of two isoforms of STAT3 directly influences the expression of a gene correlated with COVID-19 infection. 

While there is significant research in analysing specific protein isoforms, the current methods of analysing isoforms across the whole genome are limited. \acrshort{rnaseq} holds the potential of making isoform analysis accessible and accurate, without extra lab experimentation needed. The tool proposed in this paper is a digital solution which can be run on old \acrshort{rnaseq} experiments, conveying new insights.

Our chosen model organism for implementing and testing isoform prediction is Sacharomyces Cerevisiae, since it is extensively studied and many genes in more complex organisms have homologs in yeast. Scientists have often assumed that for each gene the divisions between the three regions (5'\acrshort{utr}, \acrshort{cds}, 3'\acrshort{utr}) are fixed. contrary to what was previously believed, Pelechano and Steinmetz \cite{Pelechano2013} have shown that S. cerevisiae presents an extensive level of heterogeneity in its transcriptome. In simpler terms, there is variability in the lengths of the 5’UTR and the 3’UTR (i.e. they have multiple isoforms). Thus, in the case of RNASeq experiments, it is nontrivial to know from which isoform each short read comes, especially given that these isoforms are present in various distributions under different conditions. 

It is worth mentioning that there are some alternative methods for identifying isoforms. For instance, UNAGI\cite{Alkadi2020} outperforms the Illumina pipeline when it comes to identifying novel transcripts and does slightly better when it comes to identifying isoforms. Both Alkadi et al. \cite{Alkadi2020} and Pelechano et al. \cite{Pelechano2013} require long reads of cDNA - a more complicated library preparation than the one needed for \acrshort{rnaseq}.

The study poses this question: What is the distribution of transcribed 5’UTR isoforms in an RNASeq experiment, given the findings of Pelechano and Steinmetz \cite{Pelechano2013}? To answer this, we propose developing a tool that takes any \acrshort{rnaseq} dataset together with the Pelechano annotations and outputs the relative distribution of each \acrshort{mrna} isoform. 

{\tiny [Papers for later review]} \cite{Kim2009} \cite{Lee2002} \cite{ReixachsSol2020} \cite{Thorrez2008} 

\section{Proposal - Methods}\label{methods}
\subsection{Data}
The main datasets used to run experiments are a series of eight \acrshort{rnaseq} experiments conducted as part of an ongoing study {\tiny[TODO: find a way to cite our experimets]}. These are coming from four biological conditions: two wildtype strains of S. Cerevisiae and two mutants, each coming in two replicates. In addition to these, we will use a list of the gene isoforms found by Pelechano et al. \cite{Pelechano2013}. The same study \cite{Pelechano2013} used a separate \cite{Wilkening2013}  \acrshort{rnaseq} experiment as a benchmark for their findings. Given that we are using the isoform annotations from Pelechano et al. \cite{Pelechano2013}, we will run our analysis on their benchmark dataset as well. Naturally, together with these datasets the annotated S. Cerevisiae genome will also be used. {\tiny[TODO: find a reference for it]}

{\tiny[MAYBE]}
Running some of our experiments on Ribosome footprint data could provide more insights into the positions of isoform boundaries.

\subsection{Tools}\label{tools}
{\tiny[TODO: find references for tools and libraries]}

Pre-processing the \acrshort{rnaseq} data is a critical step for our experiments. The pipeline used to filter the reads and align them to the genome was created by Ingolia et al. \cite{Ingolia2012} and it uses \textit{fastx\_toolkit} and \textit{bowtie}. The isoform identification tools are written entirely in python3 and use the following libraries to read and plot the data: \textit{pysam, pandas, pybedtools} and \textit{matplotlib}. 

\subsection{Experiments}\label{experiments}
\subsubsection{Metagene Analyisis}\label{metagene}
\acrshort{rnaseq} experiments consist of many short reads (intervals) that align with a gene (are included in a larger interval). Thus, it is important to be able to view read densities - how many reads have aligned to each individual nucleotide - over intervals of interest (often genes, or annotated isoforms). {\tiny [QUESTION: This is one of our main assumption. should it be moved somewhere else?]} For a gene with two isoforms, we expect to see a significant increase in the density right after the beginning of the shorter isoform. This is because the two isoforms overlap the same \acrshort{cds}, and since one is shorter than the other one, its 5'\acrshort{utr} should be completely included in the longer one's 5'\acrshort{utr}. Thus, any read aligned with the gene after the junction can be from either isoform, while a read aligned with the gene between the beginning of the long isoform and the junction must come from the longer isoform. 

Calculating and plotting the read densities for particular genes shows {\tiny [TODO: add a plot]} that \acrshort{rnaseq} data can be very noisy, and that the expected effect is not noticeable even for known isoform junctions. This leads to the next logical question to ask in order to validate (or invalidate) the hypothesis: would considering multiple known isoform junctions at the same time magnify the effect which is otherwise hidden by noise? In order to test this, we developed a visualisation tool that aligns and overlaps the read densities over multiple intervals. The plotted sum of these densities is called a metagene plot. To further investigate our hypothesis, we created a metagene plot of all the known isoform junctions. {\tiny [TODO: add a plot]} These preliminary results lead us to believe the hypothesis has the potential of being true. 

\subsubsection{Random generators \& Simulations}\label{rng}
In order to verify the observed effects we developed and are currently {\tiny [just for proposal]} developing a series of simulators. In a different experiment we looked at the metagene plot of all known isoform junctions. The observed increase in read density after the junction needs to be compared to a similar density calculated around randomly generated junctions.

The first attempt at random generation picked random positions from anywhere inside the genome. This revealed some complexities of random generation. First, it turns out that intergenic space has a significant number of reads, thus any junctions generated in outside annotated genes has the potential of skewing the density. Additionally, the junctions considered for the initial metagene are all residing inside the longest 5'\acrshort{utr} annotated by Pelechano et al. \cite{Pelechano2013}. Thus, a random junction generator should be genome and experiment aware. Therefore, we \textit{will} implement a random junction generator that generates random junctions informed by a set of parameters, which include a genome file, a set of annotations and whether the 
generation should follow any distribution (i.e. junctions proportional to the number of junctions identified by Pelechano et al. \cite{Pelechano2013}). With such a generator we hope to employ statistical tests to infer the likely hood of our hypothesis being true. 

One limiting factor of our datasets is that they do not include any a priori knowledge of isoform distributions. Thus, our next goal is to implement a simulation of \acrshort{rnaseq} that takes into account isoform distributions. Thus, such a simulation will start from a genome file and will generate random reads inside of all genes, proportional to weights assigned to each gene. This way, our simulation takes into account the varying levels of expression for each gene. Additionally, we hope to generate a fictional isoform database which informs the relative distributions of each gene. Subsequently, this database will serve to improve our simulation, generating an \acrshort{rnaseq} experiment for a known isoform distribution. 

\subsubsection{Isoform Predictions}\label{prediction}
The tools mentioned above should inform and provide data to subsequent isoform prediction and identification tools. 



\newacronym{rnaseq}{RNASeq}{RNA Sequencing}
\newacronym{mrna}{mRNA}{Messenger RNA}
\newacronym{cds}{CDS}{Coding Sequence}
\newacronym{utr}{UTR}{Untranslated Region}

\section{todo}
Email aitken with ingredient list for steinmetz

Prediction section of the proposal

Sharing methods

Get working on random generation
 - Make sure to go through the literature

Looking for assumptions


\printglossary[type=\acronymtype,title=Abbreviations]



\bibliography{sp}
\bibliographystyle{ieeetr}

{\tiny [TODO: Find a way to properly reference the papers in the text] }

\end{document}



